segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java

```diff
@@ -157,7 +157,7 @@ private void bootstrapOnce(EventStreamClientFactory clientFactory) {
     }
 
     private static EventStreamClientFactory createFactory(AutoScalerConfig configuration) {
-        if (configuration.isAuthEnabled()) {
+        if (configuration.isTlsEnabled()) {
```

Comment:
I'm trying to understand this change and I'm not being able to convince myself that this is correct. Could you elaborate on why you think we should limit this block to only when TLS is enabled?

Reply:
is this change needed for making the tests pass ?
it seems unrelated

Reply:
@fpj and @eolivelli 
This change is not needed for the test TlsEnabledInProcPravegaClusterTest to pass, but rather to fix the error shown below. More on this below. 

```
2019-02-13 18:51:32,820 22953 [clientInternal-1] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Establishing connection to PravegaNodeUri(endpoint=127.0.1.1, port=6000) for _system/_requeststream/0.#epoch.0, writerID: d73fec1e-706f-456b-aec3-def6e411cf3f
2019-02-13 18:51:32,820 22953 [clientInternal-1] DEBUG i.p.c.n.impl.ConnectionFactoryImpl - SslContextBuilder was set to an instance of class io.netty.handler.ssl.util.FingerprintTrustManagerFactory
2019-02-13 18:51:32,821 22954 [clientInternal-1] WARN  i.p.c.s.i.SegmentOutputStreamImpl - d73fec1e-706f-456b-aec3-def6e411cf3f Failed to connect: 
java.util.concurrent.CompletionException: java.lang.RuntimeException: javax.net.ssl.SSLException: unable to setup trustmanager
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
        at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:961)
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926)
        at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: javax.net.ssl.SSLException: unable to setup trustmanager
        at io.pravega.client.netty.impl.ConnectionFactoryImpl.establishConnection(ConnectionFactoryImpl.java:134)
        at io.pravega.client.segment.impl.SegmentOutputStreamImpl.lambda$null$1(SegmentOutputStreamImpl.java:536)
        at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:952)
        ... 9 common frames omitted
Caused by: javax.net.ssl.SSLException: unable to setup trustmanager
        at io.netty.handler.ssl.ReferenceCountedOpenSslClientContext.newSessionContext(ReferenceCountedOpenSslClientContext.java:158)
        at io.netty.handler.ssl.OpenSslClientContext.<init>(OpenSslClientContext.java:192)
        at io.netty.handler.ssl.SslContext.newClientContextInternal(SslContext.java:776)
        at io.netty.handler.ssl.SslContextBuilder.build(SslContextBuilder.java:452)
        at io.pravega.client.netty.impl.ConnectionFactoryImpl.establishConnection(ConnectionFactoryImpl.java:132)
        ... 11 common frames omitted
Caused by: java.lang.IllegalStateException: TrustManagerFactoryImpl is not initialized
        at sun.security.ssl.TrustManagerFactoryImpl.engineGetTrustManagers(TrustManagerFactoryImpl.java:100)
        at javax.net.ssl.TrustManagerFactory.getTrustManagers(TrustManagerFactory.java:285)
        at io.netty.handler.ssl.ReferenceCountedOpenSslClientContext.newSessionContext(ReferenceCountedOpenSslClientContext.java:139)
        ... 15 common frames omitted
2019-02-13 18:51:32,821 22954 [clientInt
```
TLS can be enabled without auth having been enabled, and what was happening was: the truststore and validate hostname flag were being set on the client config only when auth was enabled, which was erroneous. This was causing the error shown above. 

Now, with the change made here, if TLS is enabled, we set the truststore and validate hostname flag (line 162), and if it is not enabled we don't set it (line 168). 

https://github.com/pravega/pravega/blob/37c14774eeaa7281fa38bd1291b8e7d4c80d4447/segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java#L160-L169

With this change the logs have the following instead:

```
2019-02-13 18:56:34,805 16623 [clientInternal-3] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Establishing connection to PravegaNodeUri(endpoint=127.0.1.1, port=6000) for _system/_RGscaleGroup/0.#epoch.0, writerID: 932bcfc2-fc07-459c-a626-0f9f9ce2f3cf
2019-02-13 18:56:34,805 16623 [core-9] DEBUG i.p.s.server.writer.StorageWriter - StorageWriter[0]: Iteration[19].Finish (Elapsed 42ms).
2019-02-13 18:56:34,805 16623 [core-9] DEBUG i.p.s.server.writer.StorageWriter - StorageWriter[0]: Iteration[20].Start.
2019-02-13 18:56:34,805 16623 [core-9] DEBUG i.p.s.s.w.StorageWriterFactory$StorageWriterDataSource - WriterDataSource[0]: Read (AfterSeqNo=20, MaxCount=1000).
2019-02-13 18:56:34,805 16623 [core-9] DEBUG i.p.s.server.logs.DurableLog - DurableLog[0]: Read (AfterSequenceNumber = 20, MaxCount = 1000).
2019-02-13 18:56:34,807 16625 [epollEventLoopGroup-3-2] INFO  i.p.c.n.i.ClientConnectionInboundHandler - Connection established ChannelHandlerContext(ClientConnectionInboundHandler#0, [id: 0xe6ba3d18]) 
2019-02-13 18:56:34,807 16625 [epollEventLoopGroup-3-2] DEBUG i.p.c.n.impl.ConnectionFactoryImpl - Connect operation completed for channel:e6ba3d18, local address:/127.0.0.1:47370, remote address:127.0.1.1/127.0.1.1:6000
2019-02-13 18:56:34,808 16626 [epollEventLoopGroup-6-1] INFO  i.n.handler.logging.LoggingHandler - [id: 0x902d8f51, L:/127.0.1.1:6000] READ: [id: 0x814e873f, L:/127.0.1.1:6000 - R:/127.0.0.1:47370]
2019-02-13 18:56:34,808 16626 [epollEventLoopGroup-6-1] INFO  i.n.handler.logging.LoggingHandler - [id: 0x902d8f51, L:/127.0.1.1:6000] READ COMPLETE
2019-02-13 18:56:34,808 16626 [epollEventLoopGroup-7-4] DEBUG io.netty.handler.ssl.SslHandler - [id: 0xfe273bb8, L:/127.0.1.1:6000 - R:/127.0.0.1:47368] HANDSHAKEN: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
2019-02-13 18:56:34,811 16629 [clientInternal-4] DEBUG i.p.c.store.host.ZKHostStore - Found owning host: 127.0.1.1:6000 for containerId: 0
2019-02-13 18:56:34,812 16630 [epollEventLoopGroup-3-3] DEBUG io.netty.handler.ssl.SslHandler - [id: 0x907f5968, L:/127.0.0.1:47368 - R:127.0.1.1/127.0.1.1:6000] HANDSHAKEN: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
2019-02-13 18:56:34,813 16631 [epollEventLoopGroup-7-4] DEBUG i.p.s.s.h.h.ServerConnectionInboundHandler - Processing request: WireCommands.Hello(type=HELLO, highVersion=7, lowVersion=5)
2019-02-13 18:56:34,813 16631 [epollEventLoopGroup-7-4] INFO  i.p.s.s.h.handler.AppendProcessor - Received hello from connection: [id: 0xfe273bb8, L:/127.0.1.1:6000 - R:/127.0.0.1:47368]
2019-02-13 18:56:34,813 16631 [epollEventLoopGroup-7-4] DEBUG i.p.s.s.h.h.ServerConnectionInboundHandler - Processing request: WireCommands.ReadSegment(type=READ_SEGMENT, segment=_system/_RGscaleGroup/0.#epoch.0, offset=0, suggestedLength=1048576, delegationToken=)
2019-02-13 18:56:34,813 16631 [epollEventLoopGroup-7-4] DEBUG i.p.s.s.c.StreamSegmentContainer - SegmentContainer[0]: read [_system/_RGscaleGroup/0.#epoch.0, 0, 1048576]
2019-02-13 18:56:34,813 16631 [epollEventLoopGroup-7-4] DEBUG i.p.s.s.reading.ContainerReadIndex - ReadIndex[0]: read (StreamSegmentId = 15, Offset = 0, MaxLength = 1048576).
2019-02-13 18:56:34,813 16631 [epollEventLoopGroup-7-4] DEBUG i.p.s.s.r.StreamSegmentReadIndex - ReadIndex[0-15]: Read (Offset = 0, MaxLength = 1048576).
2019-02-13 18:56:34,814 16632 [epollEventLoopGroup-7-5] DEBUG io.netty.handler.ssl.SslHandler - [id: 0x814e873f, L:/127.0.1.1:6000 - R:/127.0.0.1:47370] HANDSHAKEN: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
2019-02-13 18:56:34,815 16633 [epollEventLoopGroup-3-3] DEBUG i.p.c.n.i.ClientConnectionInboundHandler - 127.0.1.1 processing reply: WireCommands.Hello(type=HELLO, highVersion=7, lowVersion=5)
2019-02-13 18:56:34,815 16633 [epollEventLoopGroup-3-3] INFO  i.p.s.p.netty.FailingReplyProcessor - Received hello: WireCommands.Hello(type=HELLO, highVersion=7, lowVersion=5)
2019-02-13 18:56:34,815 16633 [epollEventLoopGroup-3-3] DEBUG i.p.c.n.i.ClientConnectionInboundHandler - 127.0.1.1 processing reply: WireCommands.SegmentRead(type=SEGMENT_READ, segment=_system/_RGscaleGroup/0.#epoch.0, offset=0, atTail=true, endOfSegment=false, data=java.nio.HeapByteBuffer[pos=0 lim=340 cap=340])
2019-02-13 18:56:34,815 16633 [epollEventLoopGroup-3-2] DEBUG io.netty.handler.ssl.SslHandler - [id: 0xe6ba3d18, L:/127.0.0.1:47370 - R:127.0.1.1/127.0.1.1:6000] HANDSHAKEN: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
2019-02-13 18:56:34,816 16634 [epollEventLoopGroup-7-5] DEBUG i.p.s.s.h.h.ServerConnectionInboundHandler - Processing request: WireCommands.Hello(type=HELLO, highVersion=7, lowVersion=5)
2019-02-13 18:56:34,816 16634 [epollEventLoopGroup-7-5] INFO  i.p.s.s.h.handler.AppendProcessor - Received hello fro,,
```
Hope that clarifies.

Reply:
It's clear now.

Maybe it is worth to add a specific testcase.
From your explanation I understand that we have tests which assume that TLS = auth.

Do we have a complete matrix ?
- TLS + auth
- TLS no auth
- no TLS + auth
- no TLS no auth

Reply:
Actually, I meant that there was a bug in the main (or non-test) code that was resolved using the fix: the ClientConfig had truststore set based on a wrong condition in ``AutoScaleProcessor`` - based on ``isAuthEnabled()`` as opposed to ``isTlsEnabled()``. 

Moreover, the combinations of TLS and auth supported are accurately captured in your matrix above.  Also, there are tests for all the combinations:

|TLS Enabled?| Auth Enabled? | Test Class Name |
|:------------:|:---------------:|:-------------------:|
| No | No | InProcPravegaClusterTest|
| No  | **Yes** | PartialSecurePravegaClusterTest |
| **Yes** | **Yes** | SecurePravegaClusterTest |
| **Yes** | No |  TlsEnabledInProcPravegaClusterTest  (this was recently added)|

(On a different note, I think that the older test classes should be renamed to accurately reflect their purpose. But because that is unrelated to the issue being resolved, I didn't do that.)